<div style="text-align:left;font-size:2em"><span style="font-weight:bolder;font-size:1.25em">SP2273 | Learning Portfolio</span><br><br><span style="font-weight:bold;color:darkred">Loops (Good) Exercises</span></div>

**Exercise 1**

In [8]:
%%timeit

for number in range(10):
    if number % 2 == 0:
        continue
    square = number**2
    #print(square)


1.73 µs ± 123 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


**Exercise 2**

In [3]:
%%timeit
squares = [number**2 for number in range(10) if number % 2 != 0]
#print(squares)


2.19 µs ± 157 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


**Exercise 3**

- Exercise 1: 1.73 µs ± 123 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
- Exercise 2: 2.19 µs ± 157 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)

**Exercise 4**

In [11]:
initial_quantity = 100
decay_rate = 0.05
current_quantity = initial_quantity
years = 0

while current_quantity > initial_quantity / 2:
    current_quantity *= (1 - decay_rate)
    years += 1

print(f"It takes {years} years for the sample to reduce to half its original amount.")


It takes 14 years for the sample to reduce to half its original amount.


**Exercise 5**

In [14]:
co2_data = [
    (2000, 369.55), (2001, 371.14), (2002, 373.28), 
    (2003, 375.80), (2004, 377.52), (2005, 379.80), 
    (2006, 381.90), (2007, 383.79), (2008, 385.60), 
    (2009, 387.43), (2010, 389.90), (2011, 391.65), 
    (2012, 393.85), (2013, 396.52), (2014, 398.65),
    (2015, 400.83), (2016, 404.24), (2017, 406.55), 
    (2018, 408.52), (2019, 411.44), (2020, 414.24)
]

years_with_increase = []

for i in range(1, len(co2_data)):
    current_year, current_co2 = co2_data[i]
    previous_year, previous_co2 = co2_data[i - 1]
    change_in_concentration = current_co2 - previous_co2 
    if change_in_concentration >= 3:
        years_with_increase.append((current_year, change_in_concentration))
for year, change in years_with_increase:
    print(f"Year {year}: Change in CO2 concentration = {change:.2f} ppm")


Year 2016: Change in CO2 concentration = 3.41 ppm
